# Data Consistency Reading

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Defining path to import files
path = r'/Users/miguel/Documents/Instacart Basket Analysis'

In [3]:
# Importing orders wranged file
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# Importing products file
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Create a dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix'] = ['a','b',1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Checking for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/dr/00kxp8917rb28j9sc7qrmsqh0000gn/T/ipykernel_42550/4257676874.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [9]:
# Assigning data type to mixed column
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Finding missing values in products file
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Creating product_name missing values subset
df_nan = df_products[df_products['product_name'].isnull() == True]

In [13]:
# Checking work
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Comparing missing values to overall dataframe
df_products.shape

(49693, 5)

In [13]:
# Removing 16 missing values from products dataframe
df_products_clean = df_products[df_products['product_name'].isnull() == False]

In [14]:
# Checking work
df_products_clean.shape

(49677, 5)

In [15]:
# Looking for full duplicates in clean products dataframe
df_dups = df_products_clean[df_products_clean.duplicated()]

In [16]:
# Checking work
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
# Checking number of rows before removing duplicates
df_products_clean.shape

(49677, 5)

In [18]:
# Creating new dataframe removing duplicates
df_products_clean_no_dups = df_products_clean.drop_duplicates()

In [19]:
# Checking work
df_products_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [20]:
# 2 Finding data inconsistencies in the orders file
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

#2 Order time max should be 24, representing the 24 hours in a day.

In [21]:
# 3 Checking for mixed type data
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].map(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col)

#4 No mixed type data was found in the orders dataframe

In [22]:
# 5 Checking for missing values in orders dataframe
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

#5 There were 206,209 missing values in the 'days_since_prior_order' column. Since this is such a significant number of rows woth missing values, I believe that they are missing by design. This means there were 206,209 times where it was the first time that customers were ordering something from Instacart. 

#6 I decided to leave the missing values because they are significant for the analysis, as they represent customers making their first purchase at Instacart.

In [23]:
# 7 Checking for duplicates in the orders dataframe
df_orders_dups = df_orders[df_orders.duplicated()]

In [24]:
# 7 Check work
df_orders_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

#7 No duplicate values were found in the orders dataframe

#8 No duplicate values were found in the orders dataframe

In [25]:
df_orders.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [26]:
df_products.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))